DOSBox Pure Failing to Launch (Lakka 4.0 RPi4)

- Which version of Lakka are you using?
4.0
- What system hardware are you using?
RPi4b 4GB
- What did you do?
Launch Games
- What did you expect to happen?
Games to launch
- What happened instead?
Core exits immediately with `[ERROR] Error(s): /tmp/cores/dosbox_pure_libretro.so: cannot open shared object file: No such file or directory`

Below is the full log launching 'Archon'.

```
[INFO] [Config]: Saved new config to "/storage/.config/retroarch/retroarch.cfg".
[INFO] [Playlist]: Written to playlist file: /storage/playlists/DOS.lpl
[ERROR] Failed to open libretro core: "/tmp/cores/dosbox_pure_libretro.so"
[ERROR] Error(s): /tmp/cores/dosbox_pure_libretro.so: cannot open shared object file: No such file or directory
[INFO] [Core]: Using content: "/storage/roms/PI_EXTERNAL/retropie-mount/roms/pc/archon/archon.bat".
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [Core]: Unloading core..
[INFO] [Core]: Unloading core symbols..
[INFO] Threaded video stats: Frames pushed: 5623, Frames dropped: 0.
[INFO] RetroArch 1.10.1 (Git 4a53c67)
[INFO] === Build =======================================
[INFO] Capabilities:  ASIMD
[INFO] Built: Mar 13 2022
[INFO] Version: 1.10.1
[INFO] Git: 4a53c67
[INFO] =================================================
[INFO] [Input]: Found input driver: "udev".
[INFO] [Core]: Loading dynamic libretro core from: "/tmp/cores/dosbox_pure_libretro.so"
[ERROR] Failed to open libretro core: "/tmp/cores/dosbox_pure_libretro.so"
Error(s): /tmp/cores/dosbox_pure_libretro.so: cannot open shared object file: No such file or directory
[INFO] [Core]: Content ran for a total of: 00 hours, 00 minutes, 00 seconds.
[INFO] [Core]: Unloading core symbols..
[INFO] [Environ]: SET_PIXEL_FORMAT: RGB565.
[INFO] [Core]: Version of libretro API: 1, Compiled against API: 1
[INFO] [Audio]: Set audio input rate to: 48000.00 Hz.
[INFO] [SoftFilter]: Found plug: Normal2x Height (normal2x_height).
[INFO] [SoftFilter]: Found plug: Blargg NTSC SNES (blargg_ntsc_snes).
[INFO] [SoftFilter]: Found plug: SuperEagle (supereagle).
[INFO] [SoftFilter]: Found plug: Upscale_240x160-320x240 (upscale_240x160_320x240).
[INFO] [SoftFilter]: Found plug: Normal2x Width (normal2x_width).
[INFO] [SoftFilter]: Found plug: LQ2x (lq2x).
[INFO] [SoftFilter]: Found plug: Super2xSaI (super2xsai).
[INFO] [SoftFilter]: Found plug: Normal4x (normal4x).
[INFO] [SoftFilter]: Found plug: Darken (darken).
[INFO] [SoftFilter]: Found plug: Dot Matrix 4x (dot_matrix_4x).
[INFO] [SoftFilter]: Found plug: EPX (epx).
[INFO] [SoftFilter]: Found plug: Picoscale_256x-320x240 (picoscale_256x_320x240).
[INFO] [SoftFilter]: Found plug: Upscale1.5x (upscale_1_5x).
[INFO] [SoftFilter]: Found plug: Upscale_256x-320x240 (upscale_256x_320x240).
[INFO] [SoftFilter]: Found plug: Dot Matrix 3x (dot_matrix_3x).
[INFO] [SoftFilter]: Found plug: Grid3x (grid3x).
[INFO] [SoftFilter]: Found plug: 2xSaI (2xsai).
[INFO] [SoftFilter]: Found plug: Normal2x (normal2x).
[INFO] [SoftFilter]: Found plug: Gameboy4x (gameboy4x).
[INFO] [SoftFilter]: Found plug: Gameboy3x (gameboy3x).
[INFO] [SoftFilter]: Found plug: Phosphor2x (phosphor2x).
[INFO] [SoftFilter]: Found plug: 2xBR (2xbr).
[INFO] [SoftFilter]: Found plug: Scale2x (scale2x).
[INFO] [SoftFilter]: Found plug: Scanline2x (scanline2x).
[INFO] [SoftFilter]: Found plug: Grid2x (grid2x).
[INFO] Using 1 threads for softfilter.
[INFO] [Video]: Set video size to: fullscreen.
[INFO] [Video]: Starting threaded video driver ...
[WARN] [DRM]: Couldn't get device resources.
[WARN] [DRM]: Couldn't get device resources.
[INFO] [DRM]: Found 2 connectors.
[INFO] [DRM]: Connector 0 connected: yes
[INFO] [DRM]: Connector 0 has 38 modes.
[INFO] [DRM]: Connector 0 assigned to monitor index: #1.
[INFO] [DRM]: Connector 1 connected: no
[INFO] [DRM]: Connector 1 has 0 modes.
[INFO] [DRM]: Mode 0: (1920x1080) 1920 x 1080, 60.000000 Hz
[INFO] [DRM]: Mode 1: (1920x1080) 1920 x 1080, 59.940201 Hz
[INFO] [DRM]: Mode 2: (1920x1080i) 1920 x 1080, 30.000000 Hz
[INFO] [DRM]: Mode 3: (1920x1080i) 1920 x 1080, 29.970100 Hz
[INFO] [DRM]: Mode 4: (1920x1080) 1920 x 1080, 30.000000 Hz
[INFO] [DRM]: Mode 5: (1920x1080) 1920 x 1080, 29.970100 Hz
[INFO] [DRM]: Mode 6: (1920x1080) 1920 x 1080, 24.000000 Hz
[INFO] [DRM]: Mode 7: (1920x1080) 1920 x 1080, 23.976080 Hz
[INFO] [DRM]: Mode 8: (1680x1050) 1680 x 1050, 59.883251 Hz
[INFO] [DRM]: Mode 9: (1280x1024) 1280 x 1024, 75.024673 Hz
[INFO] [DRM]: Mode 10: (1280x1024) 1280 x 1024, 60.019741 Hz
[INFO] [DRM]: Mode 11: (1440x900) 1440 x 900, 59.901459 Hz
[INFO] [DRM]: Mode 12: (1280x960) 1280 x 960, 60.000000 Hz
[INFO] [DRM]: Mode 13: (1360x768) 1360 x 768, 60.015163 Hz
[INFO] [DRM]: Mode 14: (1280x800) 1280 x 800, 59.909546 Hz
[INFO] [DRM]: Mode 15: (1152x864) 1152 x 864, 75.000000 Hz
[INFO] [DRM]: Mode 16: (1280x720) 1280 x 720, 60.000000 Hz
[INFO] [DRM]: Mode 17: (1280x720) 1280 x 720, 60.000000 Hz
[INFO] [DRM]: Mode 18: (1280x720) 1280 x 720, 59.940201 Hz
[INFO] [DRM]: Mode 19: (1280x720) 1280 x 720, 30.000000 Hz
[INFO] [DRM]: Mode 20: (1280x720) 1280 x 720, 29.970100 Hz
[INFO] [DRM]: Mode 21: (1280x720) 1280 x 720, 24.000000 Hz
[INFO] [DRM]: Mode 22: (1280x720) 1280 x 720, 23.976162 Hz
[INFO] [DRM]: Mode 23: (1024x768) 1024 x 768, 75.028580 Hz
[INFO] [DRM]: Mode 24: (1024x768) 1024 x 768, 70.069359 Hz
[INFO] [DRM]: Mode 25: (1024x768) 1024 x 768, 60.003841 Hz
[INFO] [DRM]: Mode 26: (800x600) 800 x 600, 75.000000 Hz
[INFO] [DRM]: Mode 27: (800x600) 800 x 600, 72.187569 Hz
[INFO] [DRM]: Mode 28: (800x600) 800 x 600, 60.316540 Hz
[INFO] [DRM]: Mode 29: (720x480) 720 x 480, 60.000000 Hz
[INFO] [DRM]: Mode 30: (720x480) 720 x 480, 59.940060 Hz
[INFO] [DRM]: Mode 31: (720x480i) 720 x 480, 30.001110 Hz
[INFO] [DRM]: Mode 32: (720x480i) 720 x 480, 29.970030 Hz
[INFO] [DRM]: Mode 33: (640x480) 640 x 480, 75.000000 Hz
[INFO] [DRM]: Mode 34: (640x480) 640 x 480, 72.808800 Hz
[INFO] [DRM]: Mode 35: (640x480) 640 x 480, 60.000000 Hz
[INFO] [DRM]: Mode 36: (640x480) 640 x 480, 59.940475 Hz
[INFO] [DRM]: Mode 37: (720x400) 720 x 400, 70.081665 Hz
[INFO] [GL]: Found GL context: "kms".
[INFO] [GL]: Detecting screen resolution: 1920x1080.
[INFO] [EGL] Found EGL client version >= 1.5, trying eglGetPlatformDisplay
[INFO] [EGL]: EGL version: 1.4
[INFO] [EGL]: Current context: 0x7f8801d8f0.
[INFO] [KMS]: New FB: 1920x1080 (stride: 7680).
[INFO] [GL]: Vendor: Broadcom, Renderer: V3D 4.2.
[INFO] [GL]: Version: OpenGL ES 3.1 Mesa 22.0.0.
[INFO] [GL]: Using ARB_sync to reduce latency.
[INFO] [GL]: Using resolution 1920x1080.
[INFO] [GL]: Default shader backend found: glsl.
[INFO] [Shader driver]: Using GLSL shader backend.
[WARN] [GL]: Stock GLSL shaders will be used.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GLSL]: Found GLSL vertex shader.
[INFO] [GLSL]: Found GLSL fragment shader.
[INFO] [GLSL]: Linking GLSL program.
[INFO] [GL]: Using 4 textures.
[INFO] [GL]: Loaded 1 program(s).
[INFO] [Font]: Using font rendering backend: "freetype".
[INFO] [Video]: Graphics driver did not initialize an input driver. Attempting to pick a suitable driver.
[WARN] [udev]: Full-screen pointer won't be available.
[INFO] [udev]: Keyboard #0: "Compx 2.4G Wireless Receiver" (/dev/input/event0).
[INFO] [udev]: Keyboard #1: "Compx 2.4G Wireless Receiver Keyboard" (/dev/input/event3).
[INFO] [udev]: Keyboard #2: "Compx 2.4G Wireless Receiver Consumer Control" (/dev/input/event4).
[INFO] [udev]: Keyboard #3: "Compx 2.4G Wireless Receiver System Control" (/dev/input/event5).
[INFO] [udev]: Keyboard #4: "  Wireless Air mouse Keyboard" (/dev/input/event6).
[INFO] [udev]: Keyboard #5: "  Wireless Air mouse Keyboard Consumer Control" (/dev/input/event8).
[INFO] [udev]: Keyboard #6: "  Wireless Air mouse Keyboard System Control" (/dev/input/event9).
[INFO] [udev]: Mouse #0: "Compx 2.4G Wireless Receiver Mouse" (REL) /dev/input/event1
[INFO] [udev]: Mouse #1: "  Wireless Air mouse Keyboard Mouse" (REL) /dev/input/event7
[INFO] [udev]: Pad #0 (/dev/input/event10) supports force feedback.
[INFO] [udev]: Pad #0 (/dev/input/event10) supports 16 force feedback effects.
[INFO] [Joypad]: Found joypad driver: "udev".
[INFO] [Video]: Found display server: "null".
[WARN] [udev]: Mouse grab/ungrab feature unavailable.
[INFO] ALSA: Using floating point format.
[INFO] ALSA: Period size: 384 frames
[INFO] ALSA: Buffer size: 1536 frames
[INFO] [Display]: Found display driver: "gl".
[ERROR] Failed to open libretro core: "/tmp/cores/dosbox_pure_libretro.so"
[ERROR] Error(s): /tmp/cores/dosbox_pure_libretro.so: cannot open shared object file: No such file or directory
[ERROR] [MIDI]: snd_seq_open failed with error -2.
[ERROR] [MIDI]: Initialization failed.
[INFO] [SRAM]: SRAM will not be saved.
[INFO] [Playlist]: Loading history file: "/storage/.config/retroarch/content_history.lpl".
[INFO] [Playlist]: Loading history file: "/storage/.config/retroarch/content_music_history.lpl".
[INFO] [Playlist]: Loading history file: "/storage/.config/retroarch/content_video_history.lpl".
[INFO] [Playlist]: Loading history file: "/storage/.config/retroarch/content_image_history.lpl".
[INFO] [Playlist]: Loading favorites file: "/storage/.config/retroarch/content_favorites.lpl".
[INFO] [KMS]: New FB: 1920x1080 (stride: 7680).
[INFO] [KMS]: New FB: 1920x1080 (stride: 7680).
```

--- a/packages/lakka/libretro_cores/dosbox_pure/package.mk
+++ b/packages/lakka/libretro_cores/dosbox_pure/package.mk
@@ -7,6 +7,13 @@ PKG_DEPENDS_TARGET="toolchain"
 PKG_LONGDESC="DOSBox Pure is a fork of DOSBox, an emulator for DOS games, built for RetroArch/Libretro aiming for simplicity and ease of use."
 PKG_TOOLCHAIN="make"
 
+make_target() {
+  # remove optimization from CFLAGS, set via Makefile
+  CFLAGS="${CFLAGS//-O3/}"
+  CFLAGS="${CFLAGS//-O2/}"
+  make CXX=${CXX} CPUFLAGS="${CFLAGS}"
+}
+
 makeinstall_target() {
   mkdir -p ${INSTALL}/usr/lib/libretro
     cp -v dosbox_pure_libretro.so ${INSTALL}/usr/lib/libretro/

--- a/packages/lakka/libretro_cores/dosbox_pure/patches/dosbox_pure-Makefile-optimization_and_libs.patch
+++ b/packages/lakka/libretro_cores/dosbox_pure/patches/dosbox_pure-Makefile-optimization_and_libs.patch
@@ -1,0 +1,35 @@
+diff --git a/Makefile b/Makefile
+index 8bdbfa7..cb9674f 100644
+--- a/Makefile
++++ b/Makefile
+@@ -192,9 +194,9 @@ else
+   ifeq ($(platform),vita)
+     CFLAGS   := -DNDEBUG -O3 -fno-ident -fno-partial-inlining
+   else
+-    CFLAGS   := -DNDEBUG -O2 -fno-ident
++    CFLAGS   := -DNDEBUG -O3 -fno-ident
+   endif
+-  LDFLAGS  += -O2
++  LDFLAGS  += -O3 -fno-ident
+ endif
+ 
+ CFLAGS  += $(CPUFLAGS) -std=gnu++11 -fomit-frame-pointer -fno-exceptions -fno-non-call-exceptions -Wno-address-of-packed-member -Wno-format -Wno-switch -Wno-psabi
+@@ -209,7 +211,8 @@ CFLAGS  += -D__LIBRETRO__ -Iinclude -D_FILE_OFFSET_BITS=64
+ CFLAGS  += $(COMMONFLAGS)
+ #CFLAGS  += -fdata-sections #saves around 32 bytes on most platforms but wrongfully adds up to 60MB on msys2
+ 
+-LDFLAGS += $(CPUFLAGS) -lpthread -shared
++LDFLAGS += $(CPUFLAGS) -shared
++LDLIBS  := -lpthread
+ #LDFLAGS += -static-libstdc++ -static-libgcc #adds 1MB to output and still dynamically links against libc and libm
+ 
+ .PHONY: all clean
+@@ -234,7 +237,7 @@ ifeq ($(STATIC_LINKING), 1)
+ 	$(AR) rcs $@ $^
+ else
+ 	$(info Linking $@ ...)
+-	$(CXX) $(LDFLAGS) -o $@ $^
++	$(CXX) $(LDFLAGS) -o $@ $^ $(LDLIBS)
+ ifneq ($(BUILD),DEBUG)
+ 	@-/opt/gcw0-toolchain/usr/mipsel-gcw0-linux-uclibc/bin/strip --strip-all $@ $(PIPETONULL);true # gcw0
+ 	@-strip --strip-all $@ $(PIPETONULL);true # others

